File: Yixuan/19_01_17/sort_colors.java
class Solution {
    /**
     * @param nums: A list of integer which is 0, 1 or 2
     * @return: nothing
     * @author: Egbert Li
     * @link: http://www.lintcode.com/en/problem/sort-colors/
     * @Tag : Facebook
     */
   public void sortColors(int[] nums) {
      if (nums == null || nums.length == 0) {
         return;
      }
      int i = 0, left = 0, right = nums.length - 1;
      while (i <= right) {
         if (nums[i] == 0) {
            swap(nums, left, i);
               left++;
               i++;
            } else if (nums[i] == 1) {
               i++;
            } else {
               swap(nums, i, right);
               right--;
            }
      }
   }

   private void swap(int[] nums, int left, int right) {
      if (nums.length == 0) {
         return;
      }
      int temp = nums[left];
      nums[left] = nums[right];
        nums[right] = temp;
    }
}

File: Yixuan/19_01_17/sort_colors.py
class Solution:
    """
    @param nums: A list of integer which is 0, 1 or 2
    @return: nothing
    @link: http://www.lintcode.com/en/problem/sort-colors/
    @author: Egbert Li
    @Tag: Facebook
    """
    def sortColors(self, nums):
        if len(nums) == 0:
            return
        index, left, right = 0, 0, len(nums) - 1
        while index <= right:
            if nums[index] == 0:
                self.swap(nums, left, index)
                left += 1
                index += 1
            elif nums[index] == 1:
                index += 1
            else:
                self.swap(nums, index, right)
                right -= 1


    def swap(self, nums, left, right):
        if len(nums) == 0:
            return
        temp = nums[left]
        nums[left] = nums[right]
        nums[right] = temp
